library/Freckle/App/Kafka/Consumer/Offsets.hs

```diff
@@ -0,0 +1,54 @@
+{-# LANGUAGE TupleSections #-}
+
+module Freckle.App.Kafka.Consumer.Offsets
+  ( LastMessageByPartition
+  , newLastMessageByPartition
+  , updateLastMessageByPartition
+  , commitOffsetsSync
+  ) where
+
+import Freckle.App.Prelude
+
+import qualified Data.Map.Strict as Map
+import Kafka.Consumer
+import UnliftIO.IORef
+
+newtype LastMessageByPartition = LastMessageByPartition
+  { _unLastMessageByPartition :: IORef (Map PartitionId TopicPartition)
+  }
+
+newLastMessageByPartition :: MonadIO m => m LastMessageByPartition
+newLastMessageByPartition = LastMessageByPartition <$> newIORef Map.empty
+
+updateLastMessageByPartition
+  :: MonadIO m
+  => LastMessageByPartition
+  -> ConsumerRecord k v
+  -> m ()
+updateLastMessageByPartition (LastMessageByPartition ref) record =
+  atomicModifyIORef' ref $ (,()) . Map.insert (crPartition record) topicPartition
+ where
+  topicPartition = topicPartitionFromMessageForCommit record
+
+commitOffsetsSync
+  :: MonadIO m
+  => LastMessageByPartition
+  -> KafkaConsumer
+  -> m (Maybe KafkaError)
+commitOffsetsSync (LastMessageByPartition ref) consumer = do
+  m <- readIORef ref
+  commitPartitionsOffsets OffsetCommit consumer $ Map.elems m
```

Comment:
Do you know the distinction between `[commitAllOffsets](https://www.stackage.org/haddock/lts-22.12/hw-kafka-client-5.3.0/Kafka-Consumer.html#v:commitAllOffsets)` and `[commitPartitionsOffsets](https://www.stackage.org/haddock/lts-22.12/hw-kafka-client-5.3.0/Kafka-Consumer.html#v:commitPartitionsOffsets)`?

If we could use `commitAllOffsets` then we wouldn't need to keep track of the offset manually but I can't figure out the distinction between them.

Reply:
Seems like it plays in tandem with `enable.auto.offset.store` from https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md

Property                                 | C/P | Range           |       Default | Importance | Description              
-----------------------------------------|-----|-----------------|--------------:|------------| --------------------------
enable.auto.offset.store | C | true, false | true | high | Automatically store offset of last message provided to application. The offset store is an in-memory store of the next offset to (auto-)commit for each partition.Type: boolean
```

Reply:
`commitAllOffsets` is what I meant by "commit for all messages of the last poll" in

> This required we track the last message we processed on each commit as
we go. We could've used the "commit for all messages of the last poll"
function instead, but this would risk dropping messages if we've not
yet processed them. We seem to only poll for 1 message at a time, so
this risk is low, but I suspect we'll want to increase that and gain
free performance, so it's good to be ready.

So, that'd be why I chose not to do it that way.

Reply:
Did you consider `noAutoOffsetStore` + `storeOffsetMessage` + `commitAllOffsets`? Seems like we could still get the transactionality that we want if we `storeOffsetMessage` after consuming?

Reply:
Ah, so it looks like we could use `enable.auto.offset.store: false` and then explicitly call [`storeOffsetMessage`](https://hackage.haskell.org/package/hw-kafka-client-5.0.0/docs/Kafka-Consumer.html#v:storeOffsetMessage) only when we've processed the message. Then `commitAlloffsets` during shutdown will only commit as far as we've stored. That sounds ideal!

Reply:
Jinx!

Reply:
These techniques still imply at least once processing, right? This should just make it way less likely.

Our consumers should really be idempotent so I think that's the correct trade-off to make, I just want to make sure I am understanding correctly.

Reply:
Yes, the intent is at-least-once since we cannot tolerate drops but we can tolerate duplicate processing. I think I've achieved it?